### ISSUE TITLE
Hang in coset enumeration for a group of order 480

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#5222]
Based on an example from the forum:
```
gap> F:=FreeGroup(2);;
gap> gen:= GeneratorsOfGroup(F);;
gap> a:=gen[1];;b:=gen[2];;
gap> g:= F/[a^120, b^4, a*b*(b^3*a^41)^-1, a^2*b*(b*a^82)^-1];
<fp group on the generators [ f1, f2 ]>
```

For this group, GAP can very quickly compute the order, or an isomorphism with a pc group:
```
gap> Size(g); time;
480
5
gap> IsomorphismPcGroup(g); time;
[ f1, f2 ] -> [ f1*f2*f3^3*f4*f6*f7, f1 ]
53
```
However, `IsomorphismPermGroup(g)` has not yet terminated for me after running for a few minutes (!).

And asking for the elements results in a coset table overflow. But this only happens if I ask for `Size(g)` before. If I don't ask for the size, it completes in about 4 seconds:
```
gap> Size(g);
480
gap> Elements(g);
Error, the coset enumeration has defined more than 4096000 cosets
 at GAPROOT/lib/grpfp.gi:1234 called from
TCENUM.CosetTableFromGensAndRels( fgens, grels, fsgens ) at GAPROOT/lib/grpfp.gi:1068 called from
CosetTableFromGensAndRels( fgens, grels, fsgens ) at GAPROOT/lib/grpfp.gi:1350 called from
TryCosetTableInWholeGroup( H ) at GAPROOT/lib/grpfp.gi:1363 called from
CosetTableInWholeGroup( ImagesSet( iso, U ) ) at GAPROOT/lib/grpfp.gi:5417 called from
oper( Parent( D ), D ) at GAPROOT/lib/domain.gd:462 called from
...  at *stdin*:6
type 'return;' if you want to continue with a new limit of 8192000 cosets,
type 'quit;' if you want to quit the coset enumeration,
type 'maxlimit := 0; return;' in order to continue without a limit
brk> return;
Error, the coset enumeration has defined more than 8192000 cosets
 at GAPROOT/lib/grpfp.gi:1234 called from
TCENUM.CosetTableFromGensAndRels( fgens, grels, fsgens ) at GAPROOT/lib/grpfp.gi:1068 called from
CosetTableFromGensAndRels( fgens, grels, fsgens ) at GAPROOT/lib/grpfp.gi:1350 called from
TryCosetTableInWholeGroup( H ) at GAPROOT/lib/grpfp.gi:1363 called from
CosetTableInWholeGroup( ImagesSet( iso, U ) ) at GAPROOT/lib/grpfp.gi:5417 called from
oper( Parent( D ), D ) at GAPROOT/lib/domain.gd:462 called from
...  at *stdin*:6
type 'return;' if you want to continue with a new limit of 16384000 cosets,
type 'quit;' if you want to quit the coset enumeration,
type 'maxlimit := 0; return;' in order to continue without a limit
brk>
```

Switching to ACE for the enumeration is again fast (it is slower than it could be due to overhead for calling an external binary):
```
gap> LoadPackage("ace");
---------------------------------------------------------------------------
Loading    ACE (Advanced Coset Enumerator) 5.6.1
GAP code by Greg Gamble <Greg.Gamble@uwa.edu.au> (address for correspondence)
       Alexander Hulpke (https://www.math.colostate.edu/~hulpke)
           [uses ACE binary (C code program) version: 3.001]
C code by  George Havas (http://staff.itee.uq.edu.au/havas)
           Colin Ramsay <cram@itee.uq.edu.au>
Co-maintainer: Max Horn <horn@mathematik.uni-kl.de>

                 For help, type: ?ACE
---------------------------------------------------------------------------
true
gap> TCENUM:=ACETCENUM;;
gap> Elements(g);; time;
103
```

I can reproduce this in all GAP versions since 4.4.12, so this is not a new issue.

**Comment by fingolfin** [comment]
Maybe @hulpke or @ThomasBreuer have insights into this?

**Comment by ThomasBreuer** [comment]
I had also observed this strange behaviour, and was surprised about it:
On the one hand, GAP can quickly compute the group order and an isomorphism to a pc group (both with or without known group order).
On the other hand one gets a memory overflow when one asks `IsSolvable` in the case that the group order is known (one gets a result if the group order is not known before), and GAP has internally called `IsomorphismPermGroup` in this situation.
Calling `IsomorphismPermGroup` directly runs much longer without a memory overflow error.

**Comment by hulpke** [comment]
Basically this is an example why one should *not* work with FpGroups as a general representation for finite groups. (We have tried to make them behave better, but once in a while there are such hiccoughs.) 
Frankly, trying to calculate an automorphism group of an FpGroup (as the forum user tried) is highly naive (in the worst sense of the word) and is the "fools paradise" Neub"user's Galway proceedings article warns about.

What is happening is that a coset enumeration by <a> succeeds quickly. This is enough to establish the group's order, and is a stop criterion to use an SQ to find a faithful pc representation. If one asks for a permutation representation, GAP tries to do coset enumeration by a subgroup with trivial core, but (at least one of, the subgroup <b>) these enumerations turn out to be really hard for GAP's coset enumerator.
I have not looked at it in detail, but I suspect the reason is the large exponents (a^120, a^82) in the relators, and the lack of lookahead (which ACE has).

I know how one could fix (faithful permutation representation, not the coset enumeration) it in this example, but I'm not sure whether it would be wise to do so, as the new code would break again if the user then got encouraged to do the same for a similar group of order not 480, but 480000.

**Comment by james-d-mitchell** [comment]
I know this isn't particularly helpful, but, with `Semigroups` loaded you can do:
~~~~
gap> F:=FreeGroup(2);;
gap> gen:= GeneratorsOfGroup(F);;
gap> a:=gen[1];;b:=gen[2];;
gap> g:= F/[a^120, b^4, a*b*(b^3*a^41)^-1, a^2*b*(b*a^82)^-1];
<fp group on the generators [ f1, f2 ]>
gap> s:= Range(IsomorphismFpSemigroup(g));
<fp semigroup with 5 generators and 17 relations of length 304>
gap> IsomorphismTransformationSemigroup(s);
<fp semigroup with 5 generators and 17 relations of length 304> ->
<transformation monoid of size 480, degree 480 with 4 generators>
gap> time;
44
gap> IsomorphismPermGroup(Range(last2));
<transformation group of size 480, degree 480 with 4 generators> ->
<permutation group of size 480 with 5 generators>
gap> time;
6
~~~~

Not that I expect this to be adopted as a solution or anything, just saying...

**Comment by hulpke** [comment]
> I know this isn't particularly helpful, but, with `Semigroups` loaded you can do:

So the semigroups TC probably has something like lookahead implemented.
That *would* be worth adding to GAP's TC (which is one of the oldest untouched pieces of code we have).

**Comment by hulpke** [comment]
Resolved in #5229
